src/Microsoft.ML.Tokenizers/Normalizer/BertNormalizer.cs

```diff
@@ -0,0 +1,200 @@
+                    i += inc;
+                    continue;
+                }
+
+                if (_stripAccents && category is UnicodeCategory.NonSpacingMark or UnicodeCategory.SpacingCombiningMark)
+                {
+                    i += inc;
+                    continue;
+                }
+
+                if (_doLowerCase && category == UnicodeCategory.UppercaseLetter)
+                {
+                    int length = original.AsSpan().Slice(i, inc + 1).ToLowerInvariant(casingBuffer);
+                    Debug.Assert(length > 0);
+
+                    InsertSpan(ref buffer, ref index, casingBuffer.Slice(0, length));
+
+                    i += inc;
+                    continue;
+                }
+
+                if (_tokenizeChineseChars && IsChineseChar(codePoint))
+                {
+                    InsertChar(ref buffer, ref index, ' ');
+                    InsertChar(ref buffer, ref index, c);
+                    if (inc > 0)
+                    {
+                        InsertChar(ref buffer, ref index, original[i + 1]);
+                    }
+                    InsertChar(ref buffer, ref index, ' ');
+
+                    i += inc;
+                    continue;
+                }
+
+                InsertChar(ref buffer, ref index, c);
+                if (inc > 0)
+                {
+                    InsertChar(ref buffer, ref index, original[i + 1]);
+                }
+                i += inc;
+            }
+
+            string result = index == 0 ? string.Empty : new string(buffer, 0, index).Normalize(NormalizationForm.FormC);
+            ArrayPool<char>.Shared.Return(buffer);
+            return result;
+        }
+
+        /// <summary>
+        /// Normalize the input character span.
+        /// </summary>
+        /// <param name="original">The input character span to normalize.</param>
+        /// <returns>The normalized string.</returns>
+        public override string Normalize(ReadOnlySpan<char> original)
+        {
+            if (original.IsEmpty)
+            {
+                return string.Empty;
+            }
+
+            return Normalize(original.ToString());
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BertNormalizer"/> class.
+        /// </summary>
+        /// <param name="doLowerCase">Whether to lowercase the input.</param>
+        /// <param name="tokenizeChineseChars">Whether to tokenize Chinese characters.</param>
+        /// <param name="stripAccents">Whether to strip accents from the input.</param>
+        public BertNormalizer(bool doLowerCase, bool tokenizeChineseChars, bool stripAccents)
+        {
+            _doLowerCase = doLowerCase;
+            _tokenizeChineseChars = tokenizeChineseChars;
+            _stripAccents = stripAccents;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void InsertChar(ref char[] buffer, ref int index, char c)
```

Comment:
Generally when "Insert" is used, everything after the provided index is shifted down, but that's not happening here. Is this more of an Add or an Append rather than an Insert?